src/Microsoft.AspNetCore.Blazor.Server/Bots/RenderBatchWriter.cs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.Rendering;
+using Microsoft.AspNetCore.Blazor.RenderTree;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Blazor.Server
+{
+    // TODO: We should consider *not* having this type of infrastructure in the .Server
+    // project, but instead in some new project called .Remote or similar, since it
+    // would also be used in Electron and possibly WebWorker cases.
```

Comment:
Or if could even go in the main blazor assembly if the dependency profile is slim.

Reply:
It could. Long term we'd want to avoid `Microsoft.AspNetCore.Blazor.dll` getting big since it's used on the client, but if it's just this one class we could put that in there.

Reply:
Though I guess it should be linked out if not used, so maybe there's no disadvantage to putting it in there.